pkg/scheduler/plugins/tdm/tdm.go

```diff
@@ -0,0 +1,333 @@
+	if err != nil {
+		return err
+	}
+
+	if now.Unix() < start.Unix() || now.Unix() > end.Unix() {
+		return fmt.Errorf("current time beyond revocable zone %v:%v", rz, rzRaw)
+	}
+
+	return nil
+}
+
+func (bp *tdmPlugin) OnSessionOpen(ssn *framework.Session) {
+	klog.V(4).Infof("Enter tdm plugin ...")
+	if klog.V(4) {
+		defer func() {
+			klog.V(4).Infof("Leaving tdm plugin.")
+		}()
+	}
+
+	// tdm plugin just handle revocable node
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) error {
+		if node.RevocableZone == "" {
+			return nil
+		}
+
+		if err := bp.availableRevocableZone(node.RevocableZone); err != nil {
+			return fmt.Errorf("plugin %s predicates %w", bp.Name(), err)
+		}
+
+		klog.V(4).Infof("TDM node %v revocable zone %v:%v is active", node.Name, node.RevocableZone, bp.revocableZone[node.RevocableZone])
+
+		if !task.Preemptable {
+			msg := fmt.Sprintf("task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return fmt.Errorf("plugin %s predicates %s", bp.Name(), msg)
+		}
+
+		klog.V(4).Infof("TDM filter for Task %s/%s on node %s pass.", task.Namespace, task.Name, node.Name)
+		return nil
+	}
+
+	// tdm plugin just handle revocable node
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
+		score := 0.0
+
+		if node.RevocableZone == "" {
+			return score, nil
+		}
+
+		if err := bp.availableRevocableZone(node.RevocableZone); err != nil {
+			klog.V(4).Infof("TDM not available %s", err)
+			return score, err
+		}
+
+		if !task.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return score, nil
+		}
+
+		score = float64(v1alpha1.MaxNodeScore)
+
+		klog.V(4).Infof("TDM score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, score)
+		return score, nil
+	}
+
+	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) []*api.TaskInfo {
+		if preemptor.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is preemptable, do nothing skip", preemptor.Namespace, preemptor.Name)
+			return preemptees
+		}
+
+		var victims []*api.TaskInfo
+
+		// find preemptable tasks which appear on none revocable node
+		for _, task := range preemptees {
+			if !task.Preemptable || task.Status != api.Running {
+				continue
+			}
+
+			node, ok := ssn.Nodes[task.NodeName]
+			if !ok {
+				continue
+			}
+
+			if node.RevocableZone != "" {
+				continue
+			}
+
+			victims = append(victims, task)
+		}
+
+		active := map[string]struct{}{}
+
+		for rz := range bp.revocableZone {
+			if err := bp.availableRevocableZone(rz); err == nil {
+				active[rz] = struct{}{}
+			}
+		}
+
+		// if there are some active rz, then evict preemptable tasks from all the none revocable node
+		if len(active) > 0 {
+			victims = append(victims, bp.noneRevocableNodePreemptableTask(ssn)...)
+		}
+
+		victims = bp.maxVictims(victims)
```

Comment:
It requires to preempt pods in one node, so `maxVictims` is not suitable for this.

Reply:
Your mean in this preemptableFn we no need to call  `bp.maxVictims()`?

Reply:
Yes, it really don't need.